monitoring/procfs.go

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"os"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	ProcNetTCP  = "/proc/net/tcp"
+	ProcNetTCP6 = "/proc/net/tcp6"
+	ProcNetUDP  = "/proc/net/udp"
+	ProcNetUDP6 = "/proc/net/udp6"
+	ProcNetUnix = "/proc/net/unix"
+)
+
+type SocketState uint8
+
+const (
+	Established SocketState = 0x01
+	SynSent                 = 0x02
+	SynRecv                 = 0x03
+	FinWait1                = 0x04
+	FinWait2                = 0x05
+	TimeWait                = 0x06
+	Close                   = 0x07
+	CloseWait               = 0x08
+	LastAck                 = 0x09
+	Listen                  = 0x0A
+	Closing                 = 0x0B
+)
+
+type TCPSocket struct {
+	LocalAddress  net.TCPAddr
+	RemoteAddress net.TCPAddr
+	State         SocketState
+	TXQueue       uint
+	RXQueue       uint
+	TimerState    uint
+	TimeToTimeout uint
+	Retransmit    uint
+	UID           uint
+	Inode         uint
+}
+
+type UDPSocket struct {
+	LocalAddress  net.UDPAddr
+	RemoteAddress net.UDPAddr
+	State         SocketState
+	TXQueue       uint
+	RXQueue       uint
+	UID           uint
+	Inode         uint
+	RefCount      uint
+	Pointer       uintptr
+	Drops         uint
+}
+
+type UnixSocket struct {
+	RefCount uint
+	Protocol uint
+	Flags    uint
+	Type     uint
+	State    uint
+	Inode    uint
+	Path     string
+}
+
+func GetTCPSockets(fpath string) ([]*TCPSocket, error) {
+	var sockets []*TCPSocket
+	fp, err := os.Open(fpath)
+	defer fp.Close()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	lineScanner := bufio.NewScanner(fp)
+	lineScanner.Scan() // Drop header line
+	for lineScanner.Scan() {
+		socket, err := NewTCPSocketFromLine(lineScanner.Text())
+		if err != nil {
+			return nil, err
+		}
+		sockets = append(sockets, socket)
+	}
+	if err := lineScanner.Err(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return sockets, nil
+}
+
+func GetUDPSockets(fpath string) ([]*UDPSocket, error) {
```

Comment:
These two functions are identical except for the types - you can factor the loop into a common one and have it provide the sockets to a callback.